File: quick_sort/src/quick_sort.js
// Quick sort has average case time complexity O(nlogn), but worst
// case O(n**2).

// Not in-place. Uses O(n) memory.
Array.prototype.sort1 = function(cb) {
  if (this.length <= 1) return this;

  if (cb === undefined) {
    cb = function(x, y) {
      if(x > y) {
        return 1;
      } else {
        return 0;
      }
    };
  }

  const pivot = this[0];
  const left = [];
  const right = [];

  for (let i = 1; i < this.length; i++) {
    if (cb(this[i], pivot) === -1) {
      left.push(this[i]);
    } else {
      right.push(this[i]);
    }
  }

  return left.sort1(cb).
    concat([pivot]).
    concat(right.sort1(cb));
};

// In-place.
Array.prototype.sort2 = function(start = 0, length = this.length, cb) {
  if (this.length <= 1) return this;

  let pivotIdx = this.partition(start, length, cb);


  this.sort2(start, pivotIdx - start, cb);
  this.sort2(pivotIdx + 1, length - (pivotIdx + 1), cb);
  return this;
};

Array.prototype.partition = function(start, length, cb) {
  if (cb === undefined) {
    cb = function(x, y) {
      if(x > y) {
        return 1;
      } else {
        return 0;
      }
    };
  }

  let pivotIdx = start + 1;
  let temp;
  let idx;

  for(idx = start + 1; idx < (start + length); idx++) {
    if(cb(this[idx], this[start]) === 1) {
      pivotIdx++;
    } else {
      temp = this[idx];
      this[idx] = this[pivotIdx];
      this[pivotIdx] = temp;
    }
  }

  if (start === 0) {
    pivotIdx--;
  }

  temp = this[start];
  this[start] = this[pivotIdx];
  this[pivotIdx] = temp;
  return pivotIdx;
};

File: quick_sort/lib/quick_sort.rb
class QuickSort
  # Quick sort has average case time complexity O(nlogn), but worst
  # case O(n**2).

  # Not in-place. Uses O(n) memory.
  def self.sort1(array)
    return array if array.length <= 1

    pivot = array.first
    left = array[1..-1].select { |el| el < pivot }
    right = array[1..-1].select { |el| el >= pivot }

    QuickSort.sort1(left) + [pivot] + QuickSort.sort1(right)
  end

  # In-place.
  def self.sort2!(array, start = 0, length = array.length, &prc)
    return array if length <= 1

    # pivot = rand(start...start + length - 1)
    # array[start], array[pivot] = array[pivot], array[start]

    pivot_idx = QuickSort.partition(array, start, length, &prc)

    QuickSort.sort2!(array, start, pivot_idx - start, &prc)
    QuickSort.sort2!(array, pivot_idx + 1, length - (pivot_idx + 1), &prc)
    array
  end

  def self.partition(array, start, length, &prc)
    prc ||= Proc.new {|x, y| x <=> y}

    idx = start + 1
    pivot_idx = start + 1

    while idx < start + length
      if prc.call(array[idx], array[start]) == -1
        array[idx], array[pivot_idx] = array[pivot_idx], array[idx]
        pivot_idx += 1
      end

      idx += 1
    end
    pivot_idx -= 1
    array[start], array[pivot_idx] = array[pivot_idx], array[start]
    pivot_idx
  end
end
